tempesta_fw/sock.c

```diff
@@ -1358,11 +1362,24 @@ ss_tx_action(void)
 
 	/*
 	 * Rearm softirq for local CPU if there are more jobs to do.
-	 * ss_synchronize() is responsible for raising the softirq if there are
-	 * more jobs in the work queue or the backlog.
+	 * If all jobs are finished, and work queue and backlog are
+	 * empty, then enable IPI generation by producers (disabled
+	 * in 'ss_ipi()' handler).
+	 * ss_synchronize() is responsible for raising the softirq
+	 * if there are more jobs in the work queue or the backlog.
 	 */
-	if (!budget)
-		raise_softirq(NET_TX_SOFTIRQ);
+	if (budget) {
+		set_bit(TFW_QUEUE_B_IPI, &wq->flags);
+
+		smp_mb__after_atomic();
+
+		if (!ss_close_q_sz())
+			return;
+
+		clear_bit(TFW_QUEUE_B_IPI, &wq->flags);
```

Comment:
This is a copy of part of `tfw_wq_tasklet()` - maybe to make a macro with the queue size function as a parameter or a function accepting the queue size callback?

Reply:
Done.